src/lib/ipc/shmem/shmem.c

```diff
@@ -0,0 +1,38 @@
+#include <sys/mman.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <stdbool.h>
+
+bool shmem_unmap(void *mem, size_t size) {
+  if (munmap(mem, size) == -1) {
+    perror("munmap");
+    return(false);
+  }
+  return(true);
+}
+
+// Note: the io.* file handle passed to us is converted to a Unix
+// filehandle (FILE *) by LuaJIT
+char *shmem_grow(void *fh, void *old_mem, size_t old_size, size_t new_size) {
+  int fd;
+  void *mem;
+
+  if (old_mem != NULL) {
```

Comment:
On Thu, Jan 22, 2015 at 5:51 PM, Luke Gorrie <notifications@github.com> wrote:
>
> In src/lib/ipc/shmem/shmem.c:
>
> > +
> > +bool shmem_unmap(void *mem, size_t size) {
> > +  if (munmap(mem, size) == -1) {
> > +    perror("munmap");
> > +    return(false);
> > +  }
> > +  return(true);
> > +}
> > +
> > +// Note: the io.* file handle passed to us is converted to a Unix
> > +// filehandle (FILE *) by LuaJIT
> > +char *shmem_grow(void *fh, void *old_mem, size_t old_size, size_t new_size) {
> > +  int fd;
> > +  void *mem;
> > +
> > +  if (old_mem != NULL) {
>
> Is this function vulnerable to creating dangling pointers due to remapping memory?

Yes. That's why the shmem class keeps the locations of objects as an
offset relative
to the base address. Also see the comment of the base() method.

Now, this makes the return value of the register() method rather
useless :(  I've added
that as an afterthought to implement the semantics of your original
counter.lua code.
Clearly, I did not think this through.

> If so, could we fix that?

>
> Example:
>
> Lua code calls counter module and gets the address of a counter for repeated use.
> munmap() invalidates the old mapping and then mmap() returns a different address (for whatever reason).
> Lua code tries to use the remembered counter value and causes a SIGSEGV because it is no longer mapped.
>
> Is this a valid error scenario? (I do like the idea of counters having stable virtual addresses: then I expect they can be incremented very cheaply indeed and so we need not be shy of doing so in the fast path).

For counters, we could simply register all counters first, then use
the base() method to
create a pointer to the array of counters and use indices to access
them. The indices could
even be used as names through an "enum"-style table. This would all be
resolved at compile
time and should be as efficient as using pointers with the current code.

That wouldn't work for segments with mixed types and the return value
of register() would still
be unsafe to use.

>
> If indeed this would be a problem, could we solve it simply by skipping the munmap()? Then even if mmap() happened to choose a new address then all previous mappings would remain valid as separate mappings?

Doesn't mmap() necessarily create a new mapping when the old one isn't
munmap()-ed? Not sure.

I wonder how likely it is for mmap() to chose a new address when
explicitely asking for
a mapping to the same address. We'd still have to deal with it, even
it is unikely.